connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/internal/UplynkAdIntegration.kt

```diff
@@ -1,54 +1,181 @@
         player.addEventListener(PlayerEventTypes.TIMEUPDATE) {
             val time = it.currentTime.toDuration(DurationUnit.SECONDS)
+            //Log.d("AdScheduler", "TIMEUPDATE currentTime ${it.currentTime} time $time")
             adScheduler?.onTimeUpdate(time)
             pingScheduler?.onTimeUpdate(time)
+
+            if (state == State.PLAYING_SKIPPED_AD_BREAK) {
+                if (adScheduler?.isPlayingAd() == false) {
+                    Log.d("AdScheduler", "isPlayingAd false $time")
+                    when (uplynkConfiguration.onSeekOverAd) {
+                        SkippedAdStrategy.PLAY_NONE -> {}
+                        SkippedAdStrategy.PLAY_ALL -> {
+                            adScheduler?.getUnWatchedAdBreakOffset(seekToTime!!)
+                                ?.let { startOffset ->
+                                    snapback(startOffset)
+                                } ?: goBackToContent()
+                        }
+
+                        SkippedAdStrategy.PLAY_LAST -> {
+                            goBackToContent()
+                        }
+                    }
+                }
+            }
         }
 
         player.addEventListener(PlayerEventTypes.SEEKING) {
-            pingScheduler?.onSeeking(it.currentTime.toDuration(DurationUnit.SECONDS))
+            val playerTime = theoplayerView.player.currentTime.toDuration(DurationUnit.SECONDS)
+            val time = it.currentTime.toDuration(DurationUnit.SECONDS)
+            Log.d("AdScheduler", "TIMEUPDATE SEEKING time $time playerTime $playerTime")
+
+            if (time < playerTime) {
+                // Seeking forward
+                Log.d("AdScheduler", "TIMEUPDATE SEEKING forward")
+                handleForwardSeek(playerTime)
+            } else if (time == playerTime) {
+                // Do nothing
+            } else {
+                // Seeking backward
+                Log.d("AdScheduler", "TIMEUPDATE SEEKING backward")
+                handleBackwardSeek(playerTime)
+            }
         }
 
         player.addEventListener(PlayerEventTypes.SEEKED) {
-            pingScheduler?.onSeeked(it.currentTime.toDuration(DurationUnit.SECONDS))
+            val time = it.currentTime.toDuration(DurationUnit.SECONDS)
+            //Log.d("AdScheduler", "TIMEUPDATE SEEKED currentTime ${it.currentTime} time $time")
+            val playerTime = theoplayerView.player.currentTime.toDuration(DurationUnit.SECONDS)
+            Log.d("AdScheduler", "TIMEUPDATE SEEKED time $time playerTime $playerTime")
+
+            if (state == State.FINISHED_PLAYING_SKIPPED_AD_BREAK) {
+                state = State.PLAYING_CONTENT
+                seekToTime = null
+                pingScheduler?.onSeeked(time)
+            }
         }
 
         player.addEventListener(PlayerEventTypes.PLAY) {
             pingScheduler?.onStart(it.currentTime.toDuration(DurationUnit.SECONDS))
         }
     }
 
+    private fun handleForwardSeek(time: Duration) {
```

Comment:
While seeking forward we should only snapback if we are jumping over an unplayed ad.
The current implementation is always looking to jump into an ad regardless. Instead, we should take into consideration the current time & the seek time and only trigger the snapback logic if there's an ad in between.